src/execution/execute.ts

```diff
@@ -1294,3 +1553,444 @@ function assertEventStream(result: unknown): AsyncIterable<unknown> {
+    return { done: false, value };
+  }
+}
+
+async function executeStreamIterator(
+  initialIndex: number,
+  iterator: AsyncIterator<unknown>,
+  exeContext: ExecutionContext,
+  fieldNodes: ReadonlyArray<FieldNode>,
+  info: GraphQLResolveInfo,
+  itemType: GraphQLOutputType,
+  path?: Path,
+  label?: string,
+  parentContext?: AsyncPayloadRecord,
+): Promise<void> {
+  const subsequentPayloads = exeContext.subsequentPayloads;
+  let index = initialIndex;
+  // eslint-disable-next-line no-constant-condition
+  while (true) {
+    const fieldPath = addPath(path, index, undefined);
+    const asyncPayloadRecord = new StreamRecord({
+      label,
+      path: fieldPath,
+      parentContext,
+      iterator,
+    });
+
+    const dataPromise = executeStreamIteratorItem(
+      iterator,
+      exeContext,
+      fieldNodes,
+      info,
+      itemType,
+      asyncPayloadRecord,
+      fieldPath,
+    );
+
+    asyncPayloadRecord.addItems(
+      dataPromise
+        .then(({ value }) => value)
+        .then(
+          (value) => [value],
+          (err) => {
+            asyncPayloadRecord.errors.push(err);
+            return null;
+          },
+        ),
+    );
+    subsequentPayloads.push(asyncPayloadRecord);
+    try {
+      // eslint-disable-next-line no-await-in-loop
+      const { done } = await dataPromise;
+      if (done) {
+        break;
+      }
+    } catch (err) {
+      // do nothing, error is already handled above
+    }
+    index++;
+  }
+}
+
+function yieldSubsequentPayloads(
+  exeContext: ExecutionContext,
+  initialResult: ExecutionResult,
+): AsyncGenerator<AsyncExecutionResult, void, void> {
+  let _hasReturnedInitialResult = false;
+  let isDone = false;
+
+  async function race(): Promise<IteratorResult<AsyncExecutionResult>> {
+    if (exeContext.subsequentPayloads.length === 0) {
+      // async iterable resolver just finished and no more pending payloads
+      return {
+        value: {
+          hasNext: false,
+        },
+        done: false,
+      };
+    }
+
+    const asyncPayloadRecord: AsyncPayloadRecord = await new Promise(
+      (resolve) => {
+        exeContext.subsequentPayloads.forEach((payload) => {
+          const data = payload.getData();
+          if (isPromise(data)) {
+            // eslint-disable-next-line @typescript-eslint/no-floating-promises
+            data.then(() => {
+              resolve(payload);
+            });
+          } else {
+            resolve(payload);
+          }
+        });
+      },
+    );
+
+    if (exeContext.subsequentPayloads.length === 0) {
+      // a different call to next has exhausted all payloads
+      return { value: undefined, done: true };
+    }
+    const index = exeContext.subsequentPayloads.indexOf(asyncPayloadRecord);
+    if (index === -1) {
+      // a different call to next has consumed this payload
+      return race();
+    }
+
+    exeContext.subsequentPayloads.splice(index, 1);
```

Comment:
How about using a `Set` instead of an array for subsequentPayloads, so that this call (and the indexOf above) work in constant time? Otherwise running n defers may take O(n^2) time if they end in pessimal order.

Reply:
https://github.com/graphql/graphql-js/pull/3717